src/components/lib/vmrt/vmrt.c

```diff
@@ -0,0 +1,427 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <cos_debug.h>
+#include <cos_types.h>
+#include <llprint.h>
+#include <cos_component.h>
+#include <cos_kernel_api.h>
+#include <capmgr.h>
+
+#include <sched.h>
+#include <vmrt.h>
+#include <vmx_msr.h>
+
+void
+vmrt_dump_vcpu(struct vmrt_vm_vcpu *vcpu)
+{
+	struct vm_vcpu_shared_region *regs = vcpu->shared_region;
+
+	printc("\tRAX: %016llx\tRBX: %016llx\tRCX: %016llx\tRDX: %016llx\n",
+		regs->ax, regs->bx, regs->cx, regs->dx);
+	printc("\tR8:  %016llx\tR9:  %016llx\tR10: %016llx\tR11: %016llx\n",
+		regs->r8, regs->r9, regs->r10, regs->r11);
+	printc("\tR12: %016llx\tR13: %016llx\tR14: %016llx\tR15: %016llx\n",
+		regs->r12, regs->r13, regs->r14, regs->r15);
+	printc("\tRSI: %016llx\tRDI: %016llx\tRBP: %016llx\n",
+		regs->si, regs->di, regs->bp);
+
+	printc("\tRIP: %016llx\tRSP: %016llx\tInst Len: %llu\n", regs->ip, regs->sp, regs->inst_length);
+	printc("\tGPA: %016llx\n", regs->gpa);
+	printc("\tEFER: %016llx\t\n", regs->efer);
+	printc("\tCR0: %016llx\tCR2: %016llx\tCR4: %016llx\n", regs->cr0, regs->cr2, regs->cr4);
+}
+
+void
+vmrt_vmcs_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->vmcs_cap = capmgr_vm_vmcs_create();
+}
+
+void
+vmrt_msr_bitmap_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->msr_bitmap_cap = capmgr_vm_msr_bitmap_create();
+}
+
+void
+vmrt_lapic_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vaddr_t page;
+
+	vcpu->lapic_cap = capmgr_vm_lapic_create(&page);
+	vcpu->lapic_page = (void *)page;
+}
+
+void
+vmrt_shared_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vaddr_t page;
+
+	vcpu->shared_mem_cap = capmgr_vm_shared_region_create(&page);
+	vcpu->shared_region = (struct vm_vcpu_shared_region *)page;
+}
+
+void
+vmrt_lapic_access_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->lapic_access_cap = capmgr_vm_lapic_access_create(vcpu->vm->lapic_access_page);
+}
+
+void
+vmrt_vmcb_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->vmcb_cap = capmgr_vm_vmcb_create(vcpu->vmcs_cap, vcpu->msr_bitmap_cap, vcpu->lapic_access_cap, vcpu->lapic_cap, vcpu->shared_mem_cap, vcpu->handler_tid << 16| vcpu->vpid, 0);
+}
+
+void
+vmrt_vm_vcpu_init(struct vmrt_vm_comp *vm, u32_t vcpu_nr)
+{
+	struct vmrt_vm_vcpu *vcpu;
+	thdid_t tid, handler_tid;
+	thdcap_t cap;
+	vaddr_t resource;
+	assert(vcpu_nr < vm->num_vpu);
+
+	vcpu = &vm->vcpus[vcpu_nr];
+	assert(!vcpu->shared_region);
+
+	vcpu->vm = vm;
+
+	handler_tid = sched_thd_create(vmrt_vm_exception_handler, vcpu);
+	vcpu->handler_tid = handler_tid;
+
+	vcpu->vpid = vcpu_nr + 1;
```

Comment:
Add a comment with `/* TODO: ...*/` to notate that the vpid should be per vcpu not just for this VM, but across all VMs.

Reply:
Generally, I don't think that it is safe for this to be assigned by the VMM. If two VMMs used this policy, they might have overlapping vpids. Feels like vpids need to be assigned by the capmgr similar to (if I'm remembering correctly) ASIDs.

Would be good to discuss this.

Reply:
Oh, thanks! I didn't think into this. If there are multiple VMMs, this will not be a good solution! For now, will add a TODO there to keep in mind.